### DISCUSSION TITLE
Is it worth assigning a default value to the `scope` parameter?

### DISCUSSION THREAD

**Discussion by cedoor** [semaphore-protocol/semaphore#638] [General]
In Semaphore V4, the `generateProof` function of the `@semaphore-protocol/proof` package takes 6 parameters as input:

- the Semaphore identity instance
- the Semaphore group instance
- the message ( formerly `signal`)
- the scope ( formerly external `nullifier`)
- the maximum depth of the tree
- the paths to the snark artifacts (`.wasm`/`.zkey`)

The first four are required, the last two optional. Although there are complex applications that require special custom scopes, the scope is in most cases the group id, so that group members can only generate one valid proof within the group.

Might it make sense to also make the scope optional and assign the group id as the default value?

**Comment by 0xjei** [comment]
Thank you @cedoor for posting this. I think might be interesting to try to answer the following question: 

Should Semaphore protocol support use cases where users can generate multiple valid proofs within a group? 

For example, this would result in supporting scenarios like: voting, where users can vote multiple times until the vote is closed, or making signatures, in a group where everyone can make multiple signatures as a "form" of signaling. Please correct me if I am wrong.

**Reply by cedoor** [reply]
Yes, that's definitely something users will often need to do. I've thought about this discussion more and it could be not that worthwhile, considering all the scenarios.

Another point to consider is that `scope` is a fundamental concept that devs need to learn in order to use Semaphore in their app correctly. Making it optional could hide some complexity, but also lead devs to make mistakes due to a lack of knowledge about nullifiers.

**Comment by vplasencia** [comment]
Hey! Interesting idea. I think it's nice to keep it as it is now to support all the use cases by default. Apps using different scopes to verify many proofs inside a group are very popular too. We only have to explain the `scope` concept very clearly in the docs.
